rascal-textmate-core/src/main/rascal/lang/rascal/grammar/analyze/Symbols.rsc

```diff
@@ -0,0 +1,130 @@
+@synopsis{
+    Types and functions to analyze symbols
+}
+
+@description{
+    Note: Some functions in this module seemingly overlap with those in module
+    `lang::rascal::grammar::Lookahead` (i.e., computation of first/follow sets).
+    However, only symbols of the form `\char-class(_)` are considered terminals
+    in that module, which is too strict for the purpose of this project.
+}
+
+// TODO: The analysis of delimiters in module
+// `lang::rascal::grammar::analyze::Delimiters` can probably be rewritten (less
+// code) to use functions in this module.
+
+module lang::rascal::grammar::analyze::Symbols
+
+import Grammar;
+import ParseTree;
+import util::Maybe;
+
+import lang::rascal::grammar::Util;
+
+@synopsis{
+    Representation of a traversal direction along a list of symbols
+}
+
+data Direction   // Traverse lists of symbols (in productions)...
+    = forward()  //   - ...from left to right;
+    | backward() //   - ...from right to left.
+    ;
+
+private list[&T] reorder(list[&T] l, forward())  = l;
+private list[&T] reorder(list[&T] l, backward()) = reverse(l);
+
+@synopsis{
+    Computes the *last* set of symbol `s` in grammar `g`
+}
+
+set[Symbol] last(Grammar g, Symbol s)
+    = unmaybe(firstBySymbol(g, isTerminal, backward())[delabel(s)]);
+
+@synopsis{
+    Computes the *first* set of symbol `s` in grammar `g`
+}
+
+set[Symbol] first(Grammar g, Symbol s)
+    = unmaybe(firstBySymbol(g, isTerminal, forward())[delabel(s)]);
+
+@memo
+private map[Symbol, Maybe[set[Symbol]]] firstBySymbol(Grammar g, bool(Symbol) predicate, Direction dir) {
+    map[Symbol, Maybe[set[Symbol]]] ret
+        = (delabel(s): nothing() | s <- g.rules) // Non-terminals
+        + (delabel(s): nothing() | /prod(_, [*_, s, *_], _) := g, !isNonTerminalType(s)); // Terminals
+
+    Maybe[set[Symbol]] firstOf([])
+        = just({});
+    Maybe[set[Symbol]] firstOf([h, *t])
+        = \set: just({\empty(), *_}) := ret[delabel(h)]
+        ? union(\set, firstOf(t))
+        : ret[delabel(h)];
+
+    solve (ret) {
```

Comment:
nit: solve returns the solved value, so this can become: `return solve(ret)...`

Reply:
I played a bit with this just now, but it seems `solve` doesn't always return a value (it doesn't in the cases of this PR). I [submitted](https://github.com/usethesource/rascal/issues/2017) an issue basically to update the docs (it seems by design instead of a bug).